nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -878,6 +881,147 @@ impl DataStore {
         Ok(result)
     }
 
+    pub async fn instance_reconfigure(
+        &self,
+        opctx: &OpContext,
+        authz_instance: &authz::Instance,
+        update: InstanceUpdate,
+    ) -> Result<InstanceAndActiveVmm, Error> {
+        opctx.authorize(authz::Action::Modify, authz_instance).await?;
+
+        use crate::db::model::InstanceState;
+
+        use db::schema::disk::dsl as disk_dsl;
+        use db::schema::instance::dsl as instance_dsl;
+        use db::schema::vmm::dsl as vmm_dsl;
+
+        let err = OptionalError::new();
+        let conn = self.pool_connection_authorized(opctx).await?;
+        let (instance, vmm) = self
+            .transaction_retry_wrapper("reconfigure_instance")
+            .transaction(&conn, |conn| {
+                let err = err.clone();
+                let update = update.clone();
+                async move {
+                    // * Allow reconfiguration in NoVmm because there is no VMM
+                    //   to contend with.
+                    // * Allow reconfiguration in Failed to allow changing the
+                    //   boot disk of a failed instance and free its boot disk
+                    //   for detach.
+                    // * Allow reconfiguration in Creating because one of the
+                    //   last steps of instance creation, while the instance is
+                    //   still in Creating, is to reconfigure the instance to
+                    //   the desired boot disk.
+                    let ok_to_reconfigure_instance_states = vec![
+                        InstanceState::NoVmm,
+                        InstanceState::Failed,
+                        InstanceState::Creating,
+                    ];
+
+                    let updatable = instance_dsl::instance
+                        .filter(instance_dsl::id.eq(authz_instance.id()))
+                        .filter(
+                            instance_dsl::state
+                                .eq_any(ok_to_reconfigure_instance_states),
+                        )
+                        .filter(instance_dsl::time_deleted.is_null())
+                        .select(instance_dsl::id)
+                        .first_async::<Uuid>(&conn)
+                        .await;
+
+                    if let Err(e) = updatable {
+                        if e == diesel::NotFound {
+                            return Err(err.bail(Error::not_found_by_id(
+                                ResourceType::Instance,
+                                &authz_instance.id(),
+                            )));
```

Comment:
if the instance exists, is not deleted, but is in the wrong state for updates, this will return a `not found` instead of something more... indicative. i'll change the query to select `state` and additionally return `Error::invalid_request` if the state exists but is wrong for an update.

(it looks like `project_delete_snapshot` also has this issue)